### PR TITLE
Merge windows with same partition/order but different frames.

### PR DESCRIPTION
**Pull request by ebd2**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java [hunk 1]
```diff
@@ -79,13 +77,13 @@ public PlanNode optimize(PlanNode plan,
     }
 
     private static class Rewriter
-            extends SimplePlanRewriter<Multimap<SpecificationAndFrame, WindowNode>>
+            extends SimplePlanRewriter<Multimap<WindowNode.Specification, WindowNode>>
     {
         @Override
-        protected PlanNode visitPlan(PlanNode node, RewriteContext<Multimap<SpecificationAndFrame, WindowNode>> context)
+        protected PlanNode visitPlan(PlanNode node, RewriteContext<Multimap<WindowNode.Specification, WindowNode>> context)
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java]: Inconsistent wrapping with https://github.com/Teradata/presto/pull/317/commits/9c254786801cc3c4b7813ebb660e7ba1f9bfa1a5#diff-5537e0f0028dcf9f9a7fea8034632ee7R96.

**Comment by maciejgrzybek** [comment]
I'm not sure but maybe 5126eaf3336725efc0d416a10d4723aab5978594 should be squashed with 9c254786801cc3c4b7813ebb660e7ba1f9bfa1a5. Otherwise name of class lies.
It's easier to follow changes on PR when they are separate though.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeIdenticalWindows.java [hunk 2]
```diff
@@ -319,6 +320,54 @@ public void testNotMergeAcrossJoinBranches()
                                                 anyTree())))));
     }
 
+    @Test
+    public void testNotMergeDifferentOrdering()
+    {
+        @Language("SQL") String sql = "select " +
+                "sum(extendedprice) over (partition by suppkey order by orderkey rows between unbounded preceding and current row) sum_extendedprice_A, " +
+                "sum(quantity) over (partition by suppkey order by orderkey desc rows between UNBOUNDED preceding and CURRENT ROW) sum_quantity_C, " +
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeIdenticalWindows.java]: Keep lower / upcase consistent throughout SQL.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeIdenticalWindows.java [hunk 3]
```diff
@@ -319,6 +320,54 @@ public void testNotMergeAcrossJoinBranches()
...
+                                        functionCall("sum", windowA, false, "discount")),
+                                        anyNot(WindowNode.class)))));
+    }
+
+    @Test
+    public void testNotMergeDifferentNullOrdering()
+    {
+        @Language("SQL") String sql = "select " +
+                "sum(extendedprice) over (partition by suppkey order by orderkey rows between unbounded preceding and current row) sum_extendedprice_A, " +
+                "sum(quantity) over (partition by suppkey order by orderkey nulls first rows between UNBOUNDED preceding and CURRENT ROW) sum_quantity_C, " +
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeIdenticalWindows.java]: ditto

**Comment by ebd2** [comment]
My plan was to leave squashing the rename until just before we're going to merge with FB/master, as it's likely to cause conflicts elsewhere and is easily done automatically.

**Comment by maciejgrzybek** [comment]
@6bafbe18d5f3c0d3024213d9dcccdb53cecda26d: can you add a test with the same (order, frame) but different partition? What about a test where order key column is different but direction is the same (to discover errors where e.g. only order direction is compared but not column)?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeIdenticalWindows.java (hunk 2 above)
Comment (ebd2) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeIdenticalWindows.java]: This is kind of a mess throughout. I'll put it on a separate commit and fix all of the instances.

Comment (maciejgrzybek) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeIdenticalWindows.java]: But this test is new. You don't need to introduce such a inconsistency to fix it in other places.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java [hunk 4]
```diff
@@ -95,18 +95,18 @@ protected PlanNode visitPlan(PlanNode node, RewriteContext<Multimap<Specificatio
...
 
-            SpecificationAndFrame specificationAndFrame = new SpecificationAndFrame(node.getSpecification(), node.getFrames().iterator().next());
+            WindowNode.Specification specification = node.getSpecification();
 
             return context.rewrite(
                     node.getSource(),
-                    ImmutableListMultimap.<SpecificationAndFrame, WindowNode>builder()
-                            .put(specificationAndFrame, node) // Add the current window first so that it gets precedence in iteration order
+                    ImmutableListMultimap.<WindowNode.Specification, WindowNode>builder()
+                            .put(specification, node) // Add the current window first so that it gets precedence in iteration order
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java]: Maybe inline https://github.com/Teradata/presto/pull/317/commits/34c120d7c8a39ad46803c07a1b50a4e5a3cd97bf#diff-5537e0f0028dcf9f9a7fea8034632ee7R104 here?

**Comment by maciejgrzybek** [comment]
Merged, thanks!

**Pull request merged**